text/0779-first-class-component-templates.md

```diff
@@ -0,0 +1,1446 @@
+That is because the compilation output does *not* embed the template in the class' body in any way, but instead associates it *externally* to the class—but private class fields are only accessible within the body of the class itself, per the ECMAScript spec. While we could invest time to change the implementation to avoid this, it is not generally a problem. Because we do not interact with component class instances directly—only through the template layer—we do not actually *need* private class fields.
+
+This is a real gap, which we could address in a future RFC. Notably, it is *not specific to this proposal*, but applies to *all* proposals built on the current primitives.
+
+
+### Interop
+
+Since all existing components already work using the same low-level primitives the new system uses, strict-mode components using `<template>` can import and invoke components authored in Ember’s loose mode format. Similarly, since loose mode components are resolved to a component which is the default export of a module in the correct/conventional location on disk, components authored in strict mode with `<template>` and exported as the default export in that conventional location will be resolveable by loose mode components as well.
+
+However, other components exported as *named* exports will not be available in loose mode. This is a temporary incoherence which will be resolved as the ecosystem migrates to strict mode. As a workaround, users may choose to create reexport-only modules to allow loose-mode access. For example, given a module `app/components/named-only.js` with two named export components:
+
+```js
+export const Greet = <template>
+  <p>Hello, {{@name}}!</p>
+</template>
+
+export const Farewell = <template>
+  <p>Hello, {{@name}}!</p>
+</template>
+```
+
+A user wishing to make these available to loose mode could introduce two new modules:
+
+- `app/components/named-only/greet.js`:
+
+    ```js
+    export { Greet as default } from '../named-only';
+    ```
+
+- `app/components/named-only/farewell.js`:
+
+    ```js
+    export { Farewell as default } from '../named-only';
+    ```
+
+Then a loose-mode component could invoke them like `<NamedOnly::Greet @name="Chris">` and `<NamedOnly::Farewell @name="Krycho" />`. Note that this pattern is not at all *necessary* for migration, but may be useful.
+
+
+### The “prelude”
+
+While all values used in templates must be explicitly in scope, Ember[^glimmer-prelude] will provide some via a “prelude”.[^prelude] These are always in scope and do not need to be imported.
+
+The helpers in the prelude are:
+
+- `action`
+- `array`
+- `component`
+- `concat`
+- `debugger`
+- `each`
+- `each-in`
+- `fn`
+- `get`
+- `has-block`
+- `has-block-params`
+- `hash`
+- `helper`
+- `if`
+- `in-element`
+- `let`
+- `log`
+- `modifier`
+- `mount`
+- `mut`
+- `on`
+- `outlet`
+- `unless`
+- `yield`
+
+The modifiers in the prelude are:
+
+- `on`
+
+The components in the prelude are:
+
+- `LinkTo`
+- `Input`
+- `Textarea`
```

Comment:
Someone from Core will need to answer: is this correct? Or is the intent for these to simply be imported?

If they're meant to be imported, I will simply remove them here:

```suggestion
There are no components in the prelude.
```

Reply:
This was actually addressed via the Strict Mode Handlebars RFC. I linked there instead.